Mathlib/Analysis/SpecialFunctions/Log/NegMulLog.lean

```diff
@@ -52,18 +52,133 @@ lemma hasDerivAt_mul_log {x : ℝ} (hx : x ≠ 0) : HasDerivAt (fun x ↦ x * lo
+  by_cases xlt : L < R
+  · rw [min_eq_left_of_lt xlt, abs_eq_self.mpr hL]
+    exact Llx
+  · simp only [min_eq_right, le_of_not_lt xlt]
+    rw [abs_eq_self.mpr hR]
+    exact lt_of_le_of_lt (le_of_not_lt xlt) Llx
+
+/- helper lemma for `not_eventually_bounded_zero_mul_log` -/
+private lemma NegMulLog.one_lt_log_sub_const_of_lt_exp_sub
+    (x D : ℝ) (hx : 0 < x) (h : x < rexp (D - 2)) :
+    1 < |x.log - D| := by
+  by_cases abs_eq : x.log - D < 0
+  · rw [abs_of_neg]
+    have : x.log - D < -2 := sub_left_lt_of_lt_add ((log_lt_iff_lt_exp hx).mpr h)
+    linarith
+    exact abs_eq
+  · have h := log_lt_log hx h
+    simp only [log_exp] at h
+    rw [abs_of_pos (by linarith)]
+    linarith
+
+lemma not_eventually_bounded_derivative_zero_mul_log (D : ℝ) :
+    ¬ ∀ᶠ (x : ℝ) in 𝓝 0, |x * x.log - x * D| ≤ |x| := by
+  simp [eventually_nhdsWithin_iff, Metric.eventually_nhds_iff]
+  intro x hx
+  exists min (2⁻¹ * x) (exp (D - 2))
+  have xhalf_gt0 : 0 < 2⁻¹ * x := by norm_num [hx]
+  constructor
+  · exact abs_min_lt_of_nonneg (le_of_lt xhalf_gt0) (exp_nonneg (D - 2)) (by norm_num; linarith)
+  · have exists_pos : 0 < |min (2⁻¹ * x) (rexp (D - 2))| := by
+      simp only [one_div, abs_pos]
+      exact ne_of_gt (lt_min xhalf_gt0 (exp_pos (D - 2)))
+    simp only [← mul_sub, one_div, abs_mul]
+    apply (lt_mul_iff_one_lt_right exists_pos).mpr
+    by_cases min_l : (2⁻¹ * x) < (rexp (D - 2))
+    · rw [min_eq_left_of_lt min_l]
+      exact NegMulLog.one_lt_log_sub_const_of_lt_exp_sub (2⁻¹ * x) D xhalf_gt0 min_l
+    · simp only [min_eq_right (le_of_not_lt min_l), log_exp, sub_sub_cancel_left, abs_neg,
+        Nat.abs_ofNat, Nat.one_lt_ofNat]
+
+-- TODO inline or refactor or find home
+/- In the hypothesis, D stands for a hypothetical derivative `deriv f x0`. -/
+lemma not_DifferentiableAt_of_not_eventuallly_bounded_derivative (f : ℝ → ℝ) (x0 : ℝ)
+    (hf : ∀ D, ¬ ∀ᶠ (x : ℝ) in 𝓝 0, |f (x0 + x) - f x0 - x * D| ≤ |x|) :
+    ¬ DifferentiableAt ℝ f x0 := by
+  intro h
+  have := hasDerivAt_iff_isLittleO_nhds_zero.mp (DifferentiableAt.hasDerivAt h)
+  simp only [zero_add, log_zero, mul_zero, sub_zero, smul_eq_mul] at this
+  have := Asymptotics.IsLittleO.bound this
+  simp only [norm_mul, norm_eq_abs] at this
+  have := @this 1 (by norm_num)
+  simp only [one_mul] at this
+  have := hf (deriv f x0)
+  contradiction
+
+lemma not_DifferentiableAt_log_mul_zero :
+    ¬ DifferentiableAt ℝ (fun x ↦ x * log x) 0 := by
+  apply not_DifferentiableAt_of_not_eventuallly_bounded_derivative (fun x ↦ x * log x) 0
+  simp only [zero_add, log_zero, mul_zero, sub_zero]
+  exact not_eventually_bounded_derivative_zero_mul_log
+
+/-- Not differentiable, hence `deriv` has junk value zero. -/
+lemma deriv_mul_log_zero : deriv (fun x ↦ x * log x) 0 = 0 :=
+  deriv_zero_of_not_differentiableAt not_DifferentiableAt_log_mul_zero
+
+open Filter in
+lemma tendsto_deriv_mul_log_nhdsWithin_zero :
+    Tendsto (deriv (fun x ↦ x * log x)) (𝓝[>] 0) atBot := by
```

Comment:
This can be moved up to just below `hasDerivAt_mul_log` without changing anything. And from this, you can probably prove `lemma isLittleO_nhds_zero_id_deriv_mul_log : id =o[𝓝[>] 0] (deriv (fun x ↦ x * log x))` fairly directly, and then use that to prove non-differentiability faster.